### PR TITLE
Routing 404 fix

### PR DESCRIPTION
**Pull request by wilsonge**
This is an attempt to fix the 404 issue in the router. Note that I have done very very limited testing but this puts it out there for us to all work with. This pulls the fix from the Joomla 4 branch and implements it in the 3 branch.

Note for the 404 to be fixed any component must have a `sef_advanced` parameter. Note that if the router only uses the "new" router then this can be a hidden parameter. The parameter just needs to exist otherwise the component will have the current 404 bug (I believe that this isn't a b/c issue as no extensions currently use the new router and this parameter will be removed in Joomla 4)

### CODE REVIEW

**Comment by zero-24** [comment]
@wilsonge 

```
1) JRouterSiteTest::testParse with data set #3 ('/joomla/blog/test.json', 1, array(array('sef_suffix', null, '1')), array('', '', '', ''), array('json', 'com_test3', '45'), 'joomla/blog/test.json')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'joomla/blog/test.json'
+'joomla/blog/test'
/home/travis/build/joomla/joomla-cms/tests/unit/suites/libraries/cms/router/JRouterSiteTest.php:252
2) JRouterSiteTest::testParse with data set #4 ('/joomla/blog/test.json/', 1, array(array('sef_suffix', null, '1')), array('', '', '', ''), array('com_test3', '45'), 'joomla/blog/test.json')
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'joomla/blog/test.json'
+'joomla/blog/test'
```

File: tests/unit/suites/libraries/cms/router/JRouterSiteTest.php [hunk 1]
```diff
@@ -159,15 +159,15 @@ public function casesParse()
...
+				'expUrl'  => 'joomla/blog/test'
 			),
 			array(
 				'url'     => '/joomla/blog/test.json/',
 				'mode'    => JROUTER_MODE_SEF,
 				'map'     => array(array('sef_suffix', null, '1')),
 				'server'  => $server1,
 				'expVars' => array('option' => 'com_test3', 'Itemid' => '45'),
-				'expUrl'  => 'joomla/blog/test.json'
+				'expUrl'  => 'joomla/blog/tes'
```
Comment (zero-24) [inline tests/unit/suites/libraries/cms/router/JRouterSiteTest.php]: @wilsonge the test still fails and this looks like a typo?

btw schuld we really change the test here? Is it really an error in the tests?

Comment (wilsonge) [inline tests/unit/suites/libraries/cms/router/JRouterSiteTest.php]: Yes. I was out the house so tried to fix it on my phone :) Fixed properly now.

In terms of the tests. I honestly don't know. If the uri isn't being passed in by reference you've got to assume the URL is going to change, whether it being changed in a different way is a b/c break or not is hard to tell to be honest. In this case, honestly, it's probably the lesser evil anyhow

**Comment by Hackwar** [comment]
```
class plgSystemBreakit extends JPlugin {
	public function onAfterInitialise()
	{
		$router = JFactory::getApplication()->getRouter();
		$router->attachBuildRule(array($this, 'build'), PROCESS_AFTER);
		$router->attachParseRule(array($this, 'parse'), PROCESS_AFTER);
	}

	public function build($router, $uri)
	{
		if ($uri->hasVar('start')) {
			$uri->setPath($uri->getPath() . '/page-' . $uri->getVar('start'));
			$uri->delVar('start');
		}
	}

	public function parse($router, $uri)
	{
		$segments = explode('/', $uri->getPath());

		$last = end($segments);
		if (strpos($last, 'page-') === 0)
		{
			$uri->setVar('start', (int) str_replace($last, 'page-', ''));
		}
	}
}
```
This is just from the top of my head and not tested, but should show why the proposed code will break B/C. You don't get around deciding that behavior on a case by case basis.

**Comment by mbabker** [comment]
The 4.0 backport is happening though after all the rules are run so that shouldn't be a problem?  Otherwise in theory you've got this same issue with the plugin adding routing rules on the 4.0 branch.

**Comment by Hackwar** [comment]
The problem is not when the check is done, but that the check expects all parsed parts of the URL to be removed from the input. As you can see in my example above, that does not have to be the case.

**Comment by infograf768** [comment]
Works here when using experimental router for articles. No effect on stable router.

**Comment by wilsonge** [comment]
I'm merging this to make the conversation on routing move beyond this stupid 404 discussion. It's easy to hit the revert button if you guys have a huge issue with me merging my own code

**Pull request merged**